README.md
# Dentestimate
 


changelog.txt
Version: 4.0.1
  - Updated Bootstrap to version 5.0.0-beta2
  - Updated all outdated third party vendor libraries to their latest versions

Version: 4.0.0
  - The template does not require jQuery anymore
  - Removed jQuery and all the jQuery plugins
  - The assets/js/main.js was rewritten completely with vanilla Javascript. No more jQuery code
  - Restructured the dev version for better development experience
  - Updated the PHP Email Form to V3.0 - No jQuery dependency. Added attachment support

Version: 3.0.0
  - Initial release with Bootstrap v5.0 Beta 1

Version: 2.4.0
  - Updated Bootstrap to version 4.5.3
  - Updated all outdated third party vendor libraries to their latest versions
  - Updated the PHP Email Form to v2.3
  - Other small fixes and improvements

Version: 2.3.0
  - Updated Bootstrap to version 4.5.2
  - Updated the PHP Email Form to v2.1 
  - Updated all outdated third party vendor libraries to their latest versions
  - Other small fixes and improvements
  
  Version: 2.2.0
  - Updated Bootstrap to version 4.5.0
  - Updated the PHP Email Form library to version 2.0 with reCaptcha support
  - Added portfolio-details.html template
  - Aded inner-page.html tempalte
  - Added smooth scroll on page load with hash links in the url
  - Added navigation active state on scroll
  - Updated all outdated third party vendor libraries to their latest versions
  - Other small fixes and improvements
  
Version: 2.1.1
  - Fixed animation on scroll effect not applying after portfolio filters clicked

Version: 2.1.0
  - Added Blog page
  - Dev version: prevent updating to jQuery version 3.5.0 since it's incompatible with the current Bootstrap version 

Version: 2.0.0
  - The template was rebuilt from scratch with the latest Bootstrap version (4.4.1)
  - Added SMPTP support for the contact form script (Pro)
  - Added NodeJS NPM Development version (Pro unlimited & Membership members)
  
Version: 1.0.0
  - Initial Release


app.py
from flask import Flask, render_template, request, redirect
import os
import time
from tensorflow.keras.models import load_model
from imutils import paths
import imutils
import random
import cv2
from twilio.rest import Client
import yagmail

#imports for AI algorithm

import tensorflow.keras
from PIL import Image, ImageOps
import numpy as np

app = Flask(__name__)


@app.route('/')
def index():
    return render_template('index.html')


app.config["IMAGE_UPLOADS"] = "/Users/samarth/Downloads/Dentestimate2/static/result_img"

# Global email_array
name_array = []
email_array = []
phone_array = []


@app.route('/send', methods=['GET', 'POST'])
def send():

    if request.method == "POST":

        req = request.form

        name = req.get("name")
        name_array.append(name)
        name_array[0] = name
    

        email = req.get("email")
        email_array.append(email)
        email_array[0] = email


        phone_number = req.get("phoneNumber")
        phone_array.append(phone_number)
        phone_array[0] = phone_number



        
        return redirect(request.url)



    return render_template('index.html')


@app.route("/upload-image", methods=["GET", "POST"])
def upload_image():

    if request.method == "POST":

        if request.files:

            image = request.files["file_image"]

            image.save(os.path.join(app.config["IMAGE_UPLOADS"], "" + email_array[0] + ".png"))

            print(image)

            np.set_printoptions(suppress=True)
            model = tensorflow.keras.models.load_model('keras_model.h5')
            data = np.ndarray(shape=(1, 224, 224, 3), dtype=np.float32)

            Image.open("/Users/samarth/Downloads/Dentestimate2/static/result_img/" + email_array[0] + ".png").save("/Users/samarth/Downloads/Dentestimate2/static/result_img/" + email_array[0] + ".bmp")

            image = Image.open("/Users/samarth/Downloads/Dentestimate2/static/result_img/" + email_array[0] + ".bmp")


            size = (224, 224)
            image = ImageOps.fit(image, size, Image.ANTIALIAS)
            image_array = np.asarray(image)

            normalized_image_array = (image_array.astype(np.float32) / 127.0) - 1

            data[0] = normalized_image_array

            prediction = model.predict(data)
            
            prediction = prediction.ravel()
            prediction = prediction.tolist()
            max_value = max(prediction)

            max_index = prediction.index(max_value)

            print(prediction)
            print(max_value)
            print(max_index)

            if(max_index==0):
                return render_template('fillings.html')

            if(max_index==1):
                return render_template("crowns.html")

            if(max_index==2):
                return render_template("extractions.html")

            if(max_index==3):
                return render_template("implants.html")
            
            if(max_index==4):
                return render_template("braces.html")


            return redirect(request.url)



    return render_template("index.html")



if __name__ == "__main__":
    app.debug = True
    port = int(os.environ.get('PORT', 5000))
    app.run()


predict_fire.py
# USAGE
# python predict_fire.py

# import the necessary packages
from tensorflow.keras.models import load_model
from pyimagesearch import config
from imutils import paths
import numpy as np
import imutils
import random
import cv2
import os
from twilio.rest import Client

# load the trained model from disk
print("[INFO] loading model...")
model = load_model(config.MODEL_PATH)

# grab the paths to the fire and non-fire images, respectively
print("[INFO] predicting...")
firePaths = list(paths.list_images(config.FIRE_PATH))
nonFirePaths = list(paths.list_images(config.SAMPLE_FIRE_PATH))

# setup SMS alert system

# account_sid = os.environ["TWILIO_ACCOUNT_SID"]
# auth_token = os.environ["TWILIO_AUTH_TOKEN"]

# client = Client(account_sid, auth_token)

# media = "/Users/samarth/Downloads/keras-fire-detection/ForestFireOpenCV/sampdata/1010_nws_ocr-l-ahfire-052-1.jpg"

# combine the two image path lists, randomly shuffle them, and sample
# them
imagePaths = nonFirePaths
random.shuffle(imagePaths)
imagePaths = imagePaths[:config.SAMPLE_SIZE]

# loop over the sampled image paths
for (i, imagePath) in enumerate(imagePaths):
	# load the image and clone it
	image = cv2.imread(imagePath)
	output = image.copy()

	# resize the input image to be a fixed 128x128 pixels, ignoring
	# aspect ratio
	image = cv2.resize(image, (128, 128))
	image = image.astype("float32") / 255.0
		
	# make predictions on the image
	preds = model.predict(np.expand_dims(image, axis=0))[0]
	j = np.argmax(preds)
	label = config.CLASSES[j]

	# draw the activity on the output frame
	# text = label if label == "Non-Fire" else "Fire"

	if label == "Non-Fire":
		text = "Non-Fire"

	else:
		text = "Fire"

				
		# client.messages.create(
		# 	to="+19255773624",
		# 	from_="+12057720202",
		# 	body="Fire found!"
		# 	# media_url="file:///Users/samarth/Downloads/keras-fire-detection/ForestFireOpenCV/output/examples/0.png"

		# )


	output = imutils.resize(output, width=500)
	cv2.putText(output, text, (35, 50), cv2.FONT_HERSHEY_SIMPLEX,
		1.25, (0, 255, 0), 5)

	# write the output image to disk	 
	filename = "{}.png".format(i)
	# filename = "" + output + ".png".format(i)
	p = os.path.sep.join([config.OUTPUT_IMAGE_PATH, filename])
	cv2.imwrite(p, output)

train.py
# USAGE
# python train.py --lr-find 1
# python train.py

# set the matplotlib backend so figures can be saved in the background
import matplotlib
matplotlib.use("Agg")

# import the necessary packages
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.utils import to_categorical
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from pyimagesearch.learningratefinder import LearningRateFinder
from pyimagesearch.firedetectionnet import FireDetectionNet
from pyimagesearch import config
from imutils import paths
import matplotlib.pyplot as plt
import numpy as np
import argparse
import cv2
import sys



def load_dataset(datasetPath):
	# grab the paths to all images in our dataset directory, then
	# initialize our lists of images
	imagePaths = list(paths.list_images(datasetPath))
	data = []

	# loop over the image paths
	for imagePath in imagePaths:
		# load the image and resize it to be a fixed 128x128 pixels,
		# ignoring aspect ratio
		image = cv2.imread(imagePath)
		image = cv2.resize(image, (128, 128))

		# add the image to the data lists
		data.append(image)

	# return the data list as a NumPy array
	return np.array(data, dtype="float32")

# construct the argument parser and parse the arguments
ap = argparse.ArgumentParser()
ap.add_argument("-f", "--lr-find", type=int, default=0,
	help="whether or not to find optimal learning rate")
args = vars(ap.parse_args())

# load the fire and non-fire images
print("[INFO] loading data...")
fireData = load_dataset(config.FIRE_PATH)
nonFireData = load_dataset(config.NON_FIRE_PATH)

# construct the class labels for the data
fireLabels = np.ones((fireData.shape[0],))
nonFireLabels = np.zeros((nonFireData.shape[0],))

# stack the fire data with the non-fire data, then scale the data
# to the range [0, 1]
data = np.vstack([fireData, nonFireData])
labels = np.hstack([fireLabels, nonFireLabels])
data /= 255

# perform one-hot encoding on the labels and account for skew in the
# labeled data
labels = to_categorical(labels, num_classes=2) # 2 classes for 'fire' and 'non-fire'
classTotals = labels.sum(axis=0)
classWeight = classTotals.max() / classTotals

# construct the training and testing split
(trainX, testX, trainY, testY) = train_test_split(data, labels,
	test_size=config.TEST_SPLIT, random_state=40)
print("data", data.shape)
print("labels", labels.shape)
# X_train, X_test, y_train, y_test = train_test_split(data, labels, test_size = 0.2)





# initialize the training data augmentation object
aug = ImageDataGenerator(
	rotation_range=30, zoom_range=0.15,
	width_shift_range=0.2, height_shift_range=0.2,
	shear_range=0.15, horizontal_flip=True,
	fill_mode="nearest")




# initialize the optimizer and model
print("[INFO] compiling model...")
opt = SGD(lr=config.INIT_LR, momentum=0.9,
	decay=config.INIT_LR / config.NUM_EPOCHS)
model = FireDetectionNet.build(width=128, height=128, depth=3,
	classes=2)
model.compile(loss="binary_crossentropy", optimizer=opt,
	metrics=["accuracy"])

# check to see if we are attempting to find an optimal learning rate
# before training for the full number of epochs
if args["lr_find"] > 0:
	# initialize the learning rate finder and then train with learning
	# rates ranging from 1e-10 to 1e+1
	print("[INFO] finding learning rate...")
	lrf = LearningRateFinder(model)
	lrf.find(
		aug.flow(trainX, trainY, batch_size=config.BATCH_SIZE),
		1e-10, 1e+1,
		stepsPerEpoch=np.ceil((trainX.shape[0] / float(config.BATCH_SIZE))),
		epochs=20,
		batchSize=config.BATCH_SIZE,
		classWeight=classWeight)

	# plot the loss for the various learning rates and save the
	# resulting plot to disk
	lrf.plot_loss()
	plt.savefig(config.LRFIND_PLOT_PATH)

	# gracefully exit the script so we can adjust our learning rates
	# in the config and then train the network for our full set of
	# epochs
	print("[INFO] learning rate finder complete")
	print("[INFO] examine plot and adjust learning rates before training")
	sys.exit(0)

# train the network
print("[INFO] training network...")
H = model.fit_generator(
	aug.flow(trainX, trainY, batch_size=config.BATCH_SIZE),
	validation_data=(testX, testY),
	steps_per_epoch=trainX.shape[0] // config.BATCH_SIZE,
	epochs=config.NUM_EPOCHS,
	class_weight=classWeight,
	verbose=1)

# evaluate the network and show a classification report
print("[INFO] evaluating network...")
predictions = model.predict(testX, batch_size=config.BATCH_SIZE)
print(classification_report(testY.argmax(axis=1),
	predictions.argmax(axis=1), target_names=config.CLASSES))



# serialize the model to disk
print("[INFO] serializing network to '{}'...".format(config.MODEL_PATH))
model.save(config.MODEL_PATH)









# construct a plot that plots and saves the training history
N = np.arange(0, config.NUM_EPOCHS)
plt.style.use("ggplot")
plt.figure()
plt.plot(N, H.history["loss"], label="Train_loss")
plt.plot(N, H.history["val_loss"], label="Val_loss")
plt.plot(N, H.history["accuracy"], label="Train_acc")
plt.plot(N, H.history["val_accuracy"], label="Val_acc")
plt.title("Training Loss and Accuracy")
plt.xlabel("Epoch #")
plt.ylabel("Loss/Accuracy")
plt.legend(loc="lower left")
plt.savefig(config.TRAINING_PLOT_PATH)
